File: app/src/main/java/com/zxcx/zhizhe/ui/my/userInfo/userSafety/ChangePhoneActivity.java
package com.zxcx.zhizhe.ui.my.userInfo.userSafety;

import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONObject;
import com.zxcx.zhizhe.R;
import com.zxcx.zhizhe.mvpBase.MvpActivity;
import com.zxcx.zhizhe.ui.loginAndRegister.forget.ForgetPasswordContract;
import com.zxcx.zhizhe.ui.loginAndRegister.forget.ForgetPasswordPresenter;
import com.zxcx.zhizhe.ui.loginAndRegister.login.LoginBean;
import com.zxcx.zhizhe.ui.loginAndRegister.register.SMSCodeVerificationBean;
import com.zxcx.zhizhe.utils.SVTSConstants;
import com.zxcx.zhizhe.utils.SharedPreferencesUtil;

import java.util.regex.Pattern;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnTextChanged;
import cn.smssdk.EventHandler;
import cn.smssdk.SMSSDK;

public class ChangePhoneActivity extends MvpActivity<ForgetPasswordPresenter> implements ForgetPasswordContract.View {


    @BindView(R.id.tv_change_phone_old_phone)
    TextView mTvChangePhoneOldPhone;
    @BindView(R.id.et_change_phone_verification_code)
    EditText mEtChangePhoneVerificationCode;
    @BindView(R.id.tv_change_phone_send_verification)
    TextView mTvChangePhoneSendVerification;
    @BindView(R.id.btn_next)
    Button mBtnNext;
    @BindView(R.id.ll_change_phone_old_phone)
    LinearLayout mLlChangePhoneOldPhone;
    @BindView(R.id.et_change_phone_new_phone)
    EditText mEtChangePhoneNewPhone;
    @BindView(R.id.btn_change_phone)
    Button mBtnChangePhone;
    @BindView(R.id.ll_change_phone_new_phone)
    LinearLayout mLlChangePhoneNewPhone;
    private int count = 60;
    Handler handler = new Handler();
    String phoneRules = "^1\\d{10}$";
    Pattern phonePattern = Pattern.compile(phoneRules);
    private String phone;
    private String verifyKey;
    private String newPhone;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_change_phone);
        ButterKnife.bind(this);
        SMSSDK.registerEventHandler(new EventHandle());

        phone = SharedPreferencesUtil.getString(SVTSConstants.phone,"");
        mTvChangePhoneOldPhone.setText(phone);
    }

    @Override
    public void onBackPressed() {
        if (mLlChangePhoneNewPhone.getVisibility() == View.VISIBLE) {
            mLlChangePhoneNewPhone.setVisibility(View.GONE);
            mLlChangePhoneOldPhone.setVisibility(View.VISIBLE);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public void onDestroy() {
        handler.removeCallbacks(setDjs);
        setDjs = null;
        handler = null;
        super.onDestroy();
        SMSSDK.unregisterAllEventHandler();
    }

    @Override
    protected ForgetPasswordPresenter createPresenter() {
        return new ForgetPasswordPresenter(this);
    }

    @Override
    public void getDataSuccess(LoginBean bean) {
        SharedPreferencesUtil.saveData(SVTSConstants.phone,newPhone);
        finish();
    }

    @Override
    public void getPhoneStatusSuccess(boolean isRegistered) {
        //不需要获取手机号状态
    }

    @Override
    public void smsCodeVerificationSuccess(SMSCodeVerificationBean bean) {
        //验证码验证成功
        verifyKey = bean.getVerifyKey();
        mLlChangePhoneOldPhone.setVisibility(View.GONE);
        mLlChangePhoneNewPhone.setVisibility(View.VISIBLE);
    }

    @OnClick(R.id.iv_change_phone_close)
    public void onMIvForgetCloseClicked() {
        finish();
    }

    @OnClick(R.id.tv_change_phone_send_verification)
    public void onMTvRegisterSendVerificationClicked() {
        //发送验证码
        showLoading();
        SMSSDK.getVerificationCode("86", phone);
    }

    @OnClick(R.id.btn_next)
    public void onMBtnNextClicked() {
        //验证验证码
        mPresenter.smsCodeVerification(phone, mEtChangePhoneVerificationCode.getText().toString());
    }

    @OnClick(R.id.btn_change_phone)
    public void onMBtnRegisterClicked() {
        newPhone = mEtChangePhoneNewPhone.getText().toString();
        mPresenter.changePhone(newPhone, verifyKey);
    }

    @OnTextChanged(R.id.et_change_phone_verification_code)
    public void onRegisterNextTextChange() {
        mBtnNext.setEnabled(mEtChangePhoneVerificationCode.length() > 0);
    }

    @OnTextChanged(R.id.et_change_phone_new_phone)
    public void onRegisterPasswordTextChange() {
        mBtnChangePhone.setEnabled(checkPhone());
    }

    private boolean checkPhone() {
        if (phonePattern.matcher(mEtChangePhoneNewPhone.getText().toString()).matches()) {
            return true;
        } else {
            return false;
        }
    }

    Runnable setDjs = new Runnable() {
        @Override
        public void run() {
            mTvChangePhoneSendVerification.setText(count + " S");
            count--;

            if (count < 0) {
                handler.removeCallbacks(setDjs);
                mTvChangePhoneSendVerification.setEnabled(true);
                mTvChangePhoneSendVerification.setText(R.string.re_get_verification);

                count = 60;

                return;
            }

            handler.postDelayed(setDjs, 1000);
        }
    };

    class EventHandle extends EventHandler {
        @Override
        public void afterEvent(final int event, final int result, final Object data) {
            mActivity.runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    hideLoading();
                    if (result == SMSSDK.RESULT_COMPLETE) {
                        //回调完成
                        if (event == SMSSDK.EVENT_SUBMIT_VERIFICATION_CODE) {
                            //提交验证码成功

                        } else if (event == SMSSDK.EVENT_GET_VERIFICATION_CODE) {
                            //获取验证码成功
                            mTvChangePhoneSendVerification.setEnabled(false);
                            handler.post(setDjs);
                        } else if (event == SMSSDK.EVENT_GET_SUPPORTED_COUNTRIES) {
                            //返回支持发送验证码的国家列表
                        }
                    } else {
                        // 根据服务器返回的网络错误，给toast提示
                        try {
                            Throwable throwable = (Throwable) data;
                            throwable.printStackTrace();
                            JSONObject object = JSON.parseObject(throwable.getMessage());
                            String des = object.getString("detail");//错误描述
                            int status = object.getInteger("status");//错误代码
                            switch (status) {
                                case 457:
                                    toastShow("手机号格式错误!");
                                    break;
                                case 463:
                                case 464:
                                case 465:
                                case 477:
                                case 478:
                                    toastShow("获取验证码次数频繁，请稍后重试");
                                    break;
                                default:
                                    toastShow(des);
                                    break;
                            }
                        } catch (Exception e) {
                            //do something
                        }
                    }
                }
            });
        }
    }
}

File: app/src/main/java/com/zxcx/zhizhe/ui/my/userInfo/userSafety/ChangePhoneActivity.kt
package com.zxcx.zhizhe.ui.my.userInfo.userSafety

import android.os.Bundle
import android.os.CountDownTimer
import android.view.View
import cn.smssdk.EventHandler
import cn.smssdk.SMSSDK
import com.google.gson.JsonParser
import com.gyf.barlibrary.ImmersionBar
import com.zxcx.zhizhe.R
import com.zxcx.zhizhe.event.PhoneConfirmEvent
import com.zxcx.zhizhe.mvpBase.MvpActivity
import com.zxcx.zhizhe.ui.loginAndRegister.channelRegister.ChannelRegisterContract
import com.zxcx.zhizhe.ui.loginAndRegister.channelRegister.ChannelRegisterPresenter
import com.zxcx.zhizhe.ui.loginAndRegister.login.LoginBean
import com.zxcx.zhizhe.ui.loginAndRegister.login.PhoneConfirmDialog
import com.zxcx.zhizhe.ui.loginAndRegister.login.SMSCodeVerificationBean
import com.zxcx.zhizhe.utils.*
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.activity_change_phone.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.util.concurrent.TimeUnit
import java.util.regex.Pattern

class ChangePhoneActivity : MvpActivity<ChannelRegisterPresenter>(), ChannelRegisterContract.View {

	internal var phoneRules = "^1\\d{10}$"
	internal var phonePattern = Pattern.compile(phoneRules)
	private var phone: String? = null
	private var verifyKey: String? = null
	private var newPhone: String? = null

	val countDownTimer: CountDownTimer = object : CountDownTimer(60000, 1000) {
		override fun onFinish() {
			tv_change_phone_resend_code.isEnabled = true
			tv_change_phone_resend_code.text = "重新获取验证码"
		}

		override fun onTick(millisUntilFinished: Long) {
			val value = (millisUntilFinished / 1000).toInt()
			tv_change_phone_resend_code.text = getString(R.string.tv_login_count_down, value)
		}
	}

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_change_phone)
		EventBus.getDefault().register(this)
		SMSSDK.registerEventHandler(EventHandle())
	}

	override fun initStatusBar() {
		mImmersionBar = ImmersionBar.with(this).keyboardEnable(true)
		if (!Constants.IS_NIGHT) {
			mImmersionBar
					.statusBarColor(R.color.background)
					.statusBarDarkFont(true, 0.2f)
					.flymeOSStatusBarFontColor(R.color.text_color_1)
					.fitsSystemWindows(true)
					.keyboardEnable(true)
		} else {
			mImmersionBar
					.statusBarColor(R.color.background)
					.flymeOSStatusBarFontColor(R.color.text_color_1)
					.fitsSystemWindows(true)
					.keyboardEnable(true)
		}
		mImmersionBar.init()
	}

	override fun onBackPressed() {
		Utils.closeInputMethod(this)
		EventBus.getDefault().unregister(this)
		super.onBackPressed()
		countDownTimer.cancel()
	}

	public override fun onDestroy() {
		SMSSDK.unregisterAllEventHandler()
		super.onDestroy()
	}

	override fun createPresenter(): ChannelRegisterPresenter {
		return ChannelRegisterPresenter(this)
	}

	override fun getDataSuccess(bean: LoginBean) {
		SharedPreferencesUtil.saveData(SVTSConstants.phone, newPhone)
		finish()
	}

	fun smsCodeVerificationSuccess(bean: SMSCodeVerificationBean) {
		//验证码验证成功
		verifyKey = bean.verifyKey
		rl_change_phone_phone.visibility = View.VISIBLE
		ll_change_phone_code.visibility = View.GONE
		et_change_phone_phone.isEnabled = true
		et_change_phone_phone.setText("")
		et_change_phone_phone.setTextColor(getColorForKotlin(R.color.text_color_1))
		tv_change_phone_phone.setTextColor(getColorForKotlin(R.color.text_color_1))
		countDownTimer.onFinish()
		Utils.showInputMethod(et_change_phone_phone)
	}

	@Subscribe(threadMode = ThreadMode.MAIN)
	fun onMessageEvent(event: PhoneConfirmEvent) {
		//手机号确认成功,发送验证码
		showLoading()
		SMSSDK.getVerificationCode("86", et_change_phone_phone.text.toString())
	}

	override fun setListener() {
		iv_change_phone_close.setOnClickListener { onBackPressed() }

		tv_change_phone_phone.setOnClickListener {
			if (newPhone.isNullOrEmpty())
				return@setOnClickListener
			rl_change_phone_phone.visibility = View.VISIBLE
			ll_change_phone_code.visibility = View.GONE
			Observable.timer(300, TimeUnit.MILLISECONDS, AndroidSchedulers.mainThread()).subscribe {
				et_change_phone_phone.requestFocus()
			}
		}

		tv_change_phone_send_code.setOnClickListener {
			if (newPhone.isNullOrEmpty()) {
				showLoading()
				SMSSDK.getVerificationCode("86", et_change_phone_phone.text.toString())
			} else {
				val confirmDialog = PhoneConfirmDialog()
				val bundle = Bundle()
				bundle.putString("phone", et_change_phone_phone.text.toString())
				confirmDialog.arguments = bundle
				confirmDialog.show(mActivity.fragmentManager, "")
			}
		}

		tv_change_phone_resend_code.setOnClickListener {
			SMSSDK.getVerificationCode("86", et_change_phone_phone.text.toString())
		}

		iv_change_phone_phone_clear.setOnClickListener {
			et_change_phone_phone.setText("")
			Utils.showInputMethod(et_change_phone_phone)
		}

		//在添加监听前初始化手机号
		phone = SharedPreferencesUtil.getString(SVTSConstants.phone, "")
		et_change_phone_phone.setText(phone)
		tv_change_phone_send_code.isEnabled = true
		et_change_phone_phone.afterTextChanged {
			val isPhone = checkPhone()
			if (isPhone) {
				Utils.closeInputMethod(et_change_phone_phone)
				et_change_phone_phone.clearFocus()
				newPhone = et_change_phone_phone.text.toString()
			}
			iv_change_phone_phone_clear.visibility = if (isPhone || et_change_phone_phone.length() == 0) View.GONE else View.VISIBLE
			tv_change_phone_send_code.isEnabled = isPhone
		}

		vci_change_phone.setOnCompleteListener {
			Utils.closeInputMethod(et_change_phone_phone)
			if (newPhone.isNullOrEmpty()) {
				//mPresenter.smsCodeVerification(et_change_phone_phone.text.toString(), it)
			} else {
				SMSSDK.submitVerificationCode("86", newPhone, it)
			}
		}
	}

	private fun checkPhone(): Boolean {
		return phonePattern.matcher(et_change_phone_phone.text.toString()).matches()
	}

	internal inner class EventHandle : EventHandler() {
		override fun afterEvent(event: Int, result: Int, data: Any?) {
			mActivity.runOnUiThread {
				hideLoading()
				if (result == SMSSDK.RESULT_COMPLETE) {
					//回调完成
					when (event) {
						SMSSDK.EVENT_SUBMIT_VERIFICATION_CODE -> //提交验证码成功
							mPresenter.changePhone(newPhone, verifyKey)
						SMSSDK.EVENT_GET_VERIFICATION_CODE -> {
							//获取验证码成功
							toastShow(R.string.send_verification_code_success)
							tv_change_phone_phone.text = et_change_phone_phone.text.toString()
							rl_change_phone_phone.visibility = View.GONE
							ll_change_phone_code.visibility = View.VISIBLE
							tv_change_phone_resend_code.isEnabled = false
							Observable.timer(300, TimeUnit.MILLISECONDS, AndroidSchedulers.mainThread()).subscribe {
								vci_change_phone.focus()
							}
							countDownTimer.start()
						}
						SMSSDK.EVENT_GET_SUPPORTED_COUNTRIES -> {
							//返回支持发送验证码的国家列表
						}
					}
				} else {
					// 根据服务器返回的网络错误，给toast提示
					try {
						val throwable = data as Throwable?
						throwable!!.printStackTrace()
						val `object` = JsonParser().parse(throwable.message).asJsonObject
						val des = `object`.get("detail").asString//错误描述
						val status = `object`.get("status").asInt//错误代码
						when (status) {
							457 -> toastError("手机号格式错误!")
							463, 464, 465, 477, 478 -> toastError("获取验证码次数频繁，请稍后重试")
							else -> toastError(des)
						}
					} catch (e: Exception) {
						//do something
					}

				}
			}
		}
	}
}
